### ISSUE TITLE
StackOverflowError on readJournal.currentPersistenceIds()

### ISSUE THREAD

**Issue by gbrd** [scullxbones/akka-persistence-mongo#144]
On a "big" database (several GB) : I got a strange StackOverflowError when I try to use `currentPersistenceIds()` query.

```

val sourceOfPersistenceIds: Source[String, NotUsed] = readJournal.currentPersistenceIds()

      sourceOfPersistenceIds.runForeach { pid =>
        println("sourceOfPersistenceIds pid=" + pid)
      }

java.lang.StackOverflowError
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1230)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
[...]
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$$anonfun$$plus$plus$1.apply(Stream.scala:372)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream$$anonfun$flatMap$1.apply(Stream.scala:497)
	at scala.collection.immutable.Stream$$anonfun$flatMap$1.apply(Stream.scala:497)
	at scala.collection.immutable.Stream.append(Stream.scala:255)
	at scala.collection.immutable.Stream$$anonfun$append$1.apply(Stream.scala:255)
	at scala.collection.immutable.Stream$$anonfun$append$1.apply(Stream.scala:255)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233)
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223)
	at scala.collection.immutable.Stream.drop(Stream.scala:858)
	at scala.collection.immutable.Stream.splitAt(Stream.scala:825)
	at akka.contrib.persistence.mongodb.CurrentAllPersistenceIds.next(CasbahPersistenceReadJournaller.scala:44)
	at akka.contrib.persistence.mongodb.CurrentAllPersistenceIds.next(CasbahPersistenceReadJournaller.scala:23)
	at akka.contrib.persistence.mongodb.SyncActorPublisher$$anonfun$streaming$1.applyOrElse(MongoReadJournal.scala:367)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:497)
	at akka.contrib.persistence.mongodb.CurrentAllPersistenceIds.akka$stream$actor$ActorPublisher$$super$aroundReceive(CasbahPersistenceReadJournaller.scala:23)
	at akka.stream.actor.ActorPublisher$class.aroundReceive(ActorPublisher.scala:282)
	at akka.contrib.persistence.mongodb.CurrentAllPersistenceIds.aroundReceive(CasbahPersistenceReadJournaller.scala:23)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by scullxbones** [comment]
Hi @gbrd - yes it looks like a bug to me.  The `Stream` handling must not be quite right.

**Comment by scullxbones** [comment]
@gbrd  - if you have any ideas to help reproduce this, let me know.  I've tried very large counts of persistenceIds, with no luck.  I think I know how to fix, but would prefer to be able to confirm before cutting a release.

Thanks!

**Comment by gbrd** [comment]
Hi @scullxbones. Thanks again for your quick response !
My database is around 26GB.
We use suffixed collection names. I have around 7000 collections. (snaps+events)

What I have found for now : is does not seem to be an infinite loop (would have been very strange) : If I increase stack size (-Xss), it works "better" : 
It emits some values (I did a runWith/println on the source) and after I have some log message:

```
08:39:39.005UTC ERROR [WebcomSystem-akka.actor.default-dispatcher-19] akka://WebcomSystem/user/StreamSupervisor-1/flow-4-1-actorPublisherSource - Query failed with error code -5 and error message 'Cursor 30859064292414 not found on server 127.0.0.1:27017' on server 127.0.0.1:27017
com.mongodb.MongoCursorNotFoundException: Query failed with error code -5 and error message 'Cursor 30859064292414 not found on server 127.0.0.1:27017' on server 127.0.0.1:27017
	at com.mongodb.connection.GetMoreProtocol.execute(GetMoreProtocol.java:90) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.connection.GetMoreProtocol.execute(GetMoreProtocol.java:49) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:159) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:286) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.connection.DefaultServerConnection.getMore(DefaultServerConnection.java:251) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.operation.QueryBatchCursor.getMore(QueryBatchCursor.java:218) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.operation.QueryBatchCursor.hasNext(QueryBatchCursor.java:103) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.MongoBatchCursorAdapter.hasNext(MongoBatchCursorAdapter.java:46) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.DBCursor.hasNext(DBCursor.java:155) ~[mongo-java-driver-3.2.2.jar:na]
	at com.mongodb.casbah.MongoCursorBase$class.hasNext(MongoCursor.scala:73) ~[casbah-core_2.11-3.1.1.jar:3.1.1]
	at com.mongodb.casbah.MongoCursor.hasNext(MongoCursor.scala:487) ~[casbah-core_2.11-3.1.1.jar:3.1.1]
	at scala.collection.Iterator$class.toStream(Iterator.scala:1322) ~[na:na]
	at com.mongodb.casbah.MongoCursor.toStream(MongoCursor.scala:487) ~[casbah-core_2.11-3.1.1.jar:3.1.1]
	at scala.collection.Iterator$$anonfun$toStream$1.apply(Iterator.scala:1322) ~[na:na]
	at scala.collection.Iterator$$anonfun$toStream$1.apply(Iterator.scala:1322) ~[na:na]
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1233) ~[na:na]
	at scala.collection.immutable.Stream$Cons.tail(Stream.scala:1223) ~[na:na]
[...]
```

**Comment by scullxbones** [comment]
Hi @gbrd - it makes sense to me that increasing `-Xss` would help with the stack overflow.  I think the mongo cursor error comes from eagerly creating the cursors, but then lazily pulling the data.  Thus the cursor can easily time out (7k collections is quite a few to query across).

I'm still not having much luck (using 100k persistenceIds although with a smaller database), I had another idea.  I've published a snapshot version 1.4.1-SNAPSHOT with a fix to make the entire path lazy (was a mix of eager and lazy).  Is it possible for you to test whether the issue is resolved with that snapshot version?

**Comment by gbrd** [comment]
Thanks a lot @scullxbones : I tried 1.4.1-SNAPSHOT and it works ! (no stack overflow without increasing Xss, no cursor issue)

**Comment by scullxbones** [comment]
Great!  I'll cut `1.4.1` with that change later tonight.  Will update this ticket when it's out.

**Comment by scullxbones** [comment]
I'm going to club this together with #141 instead of releasing by itself.  Should be able to get both out this weekend.

**Comment by scullxbones** [comment]
Fixed in `1.4.1`
